docs/02-Legacy/breakdown.md

```diff
@@ -3,6 +3,302 @@ id: breakdown
+
+   - This involves creating a firebase service account, and generating a cert for it. ([reference](https://firebase.google.com/docs/admin/setup))
+
+2. **Copy `config.py.example` as `config.py` and fill in it in appropriately. `FIRESTORE_CERT` refers to the path to the firestore cert file we generated above.**
+
+### Scheduling
+
+- Start `main.py` when the RPI boots up
+
+  ```bash
+  crontab -e @reboot cd <path-to-laundro-src> && python3 runner.py main.py main`
+  ```
+
+- Check upstream repo for new updates at a interval of 1 minute
+
+  ```bash
+  crontab -e * * * * * cd path/to/laundro && git pull && git log --pretty=oneline -1 | python3 flagger.py
+  ```
+
+## How it Works
+
+### `main.py`
+
+**This is where the main control loop resides and the entrypoint of the program.**
+
+:::info
+Also provides a control loop (`local_main`) that updates pin values (to a `.csv`) without connection to firebase. This is called when an error is thrown when triggering the actual control loop.
+:::
+
+```python
+def main():
+    # Initiates a logger
+    init_logger(config.LOGDIR)
+    log = logging.getLogger()
+    log.info('main script started.')
+
+    # Set up Firestore connection
+    firestore = FirestoreManager(config.FIRESTORE_CERT, config.PI_ID)
+
+    # Get the list of pins
+    # Only grabs the pins that are allocated to config.PI_ID (more on this below)
+    pins = get_pins(config.PI_ID)
+    flag = flagger.Flag(flagger.flag)
+
+    # Creates firestore documents for the pins and rpi, if it doesn't exist already
+    firestore.init_pins(map(lambda p: p.id, pins))
+    firestore.init_pi()
+
+    # Grab last recorded pin readings from firestore
+    prev_on = {} # { <pin_id: int> :  is_on: bool }
+    washer_ids = firestore.get_washing_machine_pin_ids()
+    for p in pins:
+        on = p.is_on()
+        prev_on[p.id] = on
+
+        # For each pin, get the previous time when it was updated from firebase
+        current = firestore.get_pin_data(p.id)
+
+        now = sg_time_now()
+        timediff = now - current['timeChanged']
+
+        '''
+        If the the status of the pin is unchanged for a period under 30 min (for washers) / 45 min (for dryers) we not modify the pin record
+
+        Intuition:
+        - To update timeChanged of the pin status upon script start
+        - However since we are not certain of what is the exact timestamp this change happen, timeChangedCertain = False
+
+        Possible Concern:
+        - Can we actually trust the clocks of the RPI?
+        '''
+        if on == current['on'] and p.id in washer_ids and timediff <= dt.timedelta(minutes=30) or \
+           timediff <= dt.timedelta(minutes=45):
+            continue
+        else:
+            firestore.update_pin(p.id, on, sg_time_now(), timeChangedCertain = False)
+
+    '''
+    Assumes that each run of the loop takes around 25 seconds
+
+    Observation:
+    Runtime of one iteration = 2secs * 9pins + 1 secs sleep + 2 secs firebase/logs ~= 21secs
```

Comment:
is `2 secs firebase/logs` an estimation or was it stated somewhere in their repo?

Reply:
its estimation, i think rough figure was from logs file?

Reply:
Is it possible to just put a snippet of the logs that shows it takes around 2 secs? You can find log dump at the top of the page in the current head of the branch